chore(repo-server): unify semver resolution in new versions subpackage

PaulSonOfLars opened the pull request:
Unify semver resolution across helm/OCI/git into a single package, to ensure we don't end up with mismatched logic for different services 

Follow up to #20096, as requested [here](https://github.com/argoproj/argo-cd/pull/20096#pullrequestreview-2335883449). cc @blakepettersson 

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

blakepettersson commented on util/helm/client.go:
```diff
@@ -414,7 +413,11 @@ func getIndexURL(rawURL string) (string, error) {
 }
 
-func (c *nativeHelmChart) GetTags(chart string, noCache bool) (*TagsList, error) {
+type Entries struct {
```
Could we simplify this to just be a simple string array instead of a separate type?

PaulSonOfLars replied:
So, I had a look into it, and this specific structure is used on L496 to unmarshal a JSON response from helm.

We could make it a simple array, and just inline the struct if that would be preferred - would definitely be clearer (and avoid reusing a struct where it shouldn't be)

blakepettersson commented on util/versions/version.go:
```diff
@@ -1,4 +1,4 @@
+package versions
 
 import "github.com/Masterminds/semver/v3"
 
```
I reviewed this in my OCI branch, and I think something that would be useful would be a `IsConstraint` function.

```
func IsConstraint(text string) bool {
	_, err := semver.NewConstraint(text)
	return err == nil
}
```

PaulSonOfLars replied:
That sounds good - but for the sake of our logic, I'd recommend ensuring that a constraint can never be a valid version (which is the root of this whole issue!)

```
func IsConstraint(text string) bool {
	_, err := semver.NewConstraint(text)
	return !IsVersion(text) && err == nil
}
```

blakepettersson reviewed (commented):
This is looking good so far! Just a few minor comments

blakepettersson commented:
I think you also need to run `make mockgen`

blakepettersson commented on util/helm/index.go:
```diff
@@ -1,59 +1,17 @@
-
 type Index struct {
-	Entries map[string]Entries
+	Entries map[string][]string
```
I was looking at this and I think we should keep the `Index` code as-is.

blakepettersson commented on util/helm/index_test.go:
```diff
@@ -3,80 +3,34 @@ package helm
 
 var index = Index{
-	Entries: map[string]Entries{
+	Entries: map[string][]string{
```
The same applies here

blakepettersson commented on reposerver/repository/repository.go:
```diff
@@ -2419,39 +2419,34 @@ func (s *Service) newClientResolveRevision(repo *v1alpha1.Repository, revision s
+		if err != nil {
+			return nil, "", err
+		}
+		tags, err = index.GetTags(chart)
```
Once we get the entries changes back then we would need to convert the entries back into tags.

```suggestion
		entries, err := index.GetEntries(chart)
		for _, entry := range entries {
			tags = append(tags, entry.Version)
		}
```

blakepettersson commented on reposerver/repository/repository_test.go:
```diff
@@ -122,9 +122,9 @@ func newServiceWithMocks(t *testing.T, root string, signed bool) (*Service, *git
-		helmClient.On("GetIndex", mock.AnythingOfType("bool"), mock.Anything).Return(&helm.Index{Entries: map[string]helm.Entries{
-			chart:    {{Version: "1.0.0"}, {Version: version}},
-			oobChart: {{Version: "1.0.0"}, {Version: version}},
+		helmClient.On("GetIndex", mock.AnythingOfType("bool"), mock.Anything).Return(&helm.Index{Entries: map[string][]string{
```
This should also be reverted to the previous state

blakepettersson commented on util/helm/client_test.go:
```diff
@@ -312,14 +310,12 @@ func TestGetTagsFromURLEnvironmentAuthentication(t *testing.T) {
-				"2.8.0-prerelease_build",
-				"2.8.0-prerelease.1_build.1234",
-			},
+		responseTags := []string{
+			"2.8.0",
+			"2.8.0-prerelease",
+			"2.8.0_build",
+			"2.8.0-prerelease_build",
+			"2.8.0-prerelease.1_build.1234",
 		}
```
For testing purposes we can add a private `tagsList` type in the test itself and then revert to almost the same thing as before

```suggestion
		responseTags := tagsList{
			Tags: []string{
				"2.8.0",
				"2.8.0-prerelease",
				"2.8.0_build",
				"2.8.0-prerelease_build",
				"2.8.0-prerelease.1_build.1234",
			},
		}
```

blakepettersson commented on util/helm/client_test.go:
```diff
@@ -231,14 +231,12 @@ func TestGetTagsFromURLPrivateRepoAuthentication(t *testing.T) {
-				"2.8.0-prerelease_build",
-				"2.8.0-prerelease.1_build.1234",
-			},
+		responseTags := []string{
+			"2.8.0",
+			"2.8.0-prerelease",
+			"2.8.0_build",
+			"2.8.0-prerelease_build",
+			"2.8.0-prerelease.1_build.1234",
 		}
```
```suggestion
		responseTags := tagsList{
			Tags: []string{
				"2.8.0",
				"2.8.0-prerelease",
				"2.8.0_build",
				"2.8.0-prerelease_build",
				"2.8.0-prerelease.1_build.1234",
			},
		}
```

blakepettersson commented on util/helm/index.go:
```diff
@@ -1,59 +1,17 @@
 
-func (i *Index) GetEntries(chart string) (Entries, error) {
-	entries, ok := i.Entries[chart]
+func (i *Index) GetTags(chart string) ([]string, error) {
```
This means that for this case we would need to revert `GetTags` back to `GetEntries`.

PaulSonOfLars commented:
@blakepettersson thanks for the review! 
I've reverted things back to entries and added in a private struct for the tags methods. Also fixed some tests which were failing due to new error messages. Let me know if you'd like me to poke at anything else!

blakepettersson commented on util/versions/version.go:
```diff
@@ -1,8 +1,13 @@
 	_, err := semver.NewVersion(text)
 	return err == nil
 }
+
+func IsConstraint(text string) bool {
+	_, err := semver.NewConstraint(text)
+	return !IsVersion(text) && err == nil
+}
```
Actually, I took another stab at this and I think this can be removed. 
```suggestion
```

blakepettersson reviewed (approved):
One minor thing, otherwise this looks great. Thanks!

Spazzy757 commented:
Hey!

Is there any update on this? We have a pending bug that we would like fixed, but we are waiting on this PR :smile: 

https://github.com/argoproj/argo-cd/pull/20862

gh0st3e commented:
Hey, how it is going? I really appreciate your work because i really need this feature. Hope you don't forgot about it.

jose-arcatar commented:
we need these feature as well

jackalbsd commented:
Hello guys.
Could you please share the possible time when this PR can be merged?

We depend on the fix (which is in this PR) for this issue - https://github.com/argoproj/argo-cd/issues/21818
At the moment we are blocked. 
We cannot upgrade ArgoCD to 2.14 because of the issue,  at the same time we cannot upgrade k8s to 1.31 because ArgoCD 2.13 is not supported in k8s 1.31.

Appreciate for your help.
Thanks

Pacobart commented:
Also blocked on this feature. Is there an ETA?

PaulSonOfLars commented on reposerver/repository/repository.go:
```diff
@@ -2414,40 +2414,32 @@ func (s *Service) newClientResolveRevision(repo *v1alpha1.Repository, revision s
 func (s *Service) newHelmClientResolveRevision(repo *v1alpha1.Repository, revision string, chart string, noRevisionCache bool) (helm.Client, string, error) {
 	enableOCI := repo.EnableOCI || helm.IsHelmOciRepo(repo.Repo)
 	helmClient := s.newHelmClient(repo.Repo, repo.GetHelmCreds(), enableOCI, repo.Proxy, repo.NoProxy, helm.WithIndexCache(s.cache), helm.WithChartPaths(s.chartPaths))
-	if helm.IsVersion(revision) {
-		return helmClient, revision, nil
-	}
-	constraints, err := semver.NewConstraint(revision)
-	if err != nil {
-		return nil, "", fmt.Errorf("invalid revision '%s': %w", revision, err)
-	}
```
This is now integrated as part of versions.MaxVersion, improving version validation and ensuring we don't accidentally return invalid data. It also consolidates git vs helm logic.

PaulSonOfLars commented on util/helm/index_test.go:
```diff
@@ -30,53 +29,8 @@ func TestIndex_GetEntries(t *testing.T) {
-	})
-}
-
-func TestEntries_MaxVersion(t *testing.T) {
```
This is moved to the `versions` package

PaulSonOfLars commented on util/git/client.go:
```diff
@@ -701,59 +710,6 @@ func (m *nativeGitClient) lsRemote(revision string) (string, error) {
 	return "", fmt.Errorf("unable to resolve '%s' to a commit SHA", revision)
 }
 
-// resolveSemverRevision is a part of the lsRemote method workflow.
```
This is all part of the `versions` package now

PaulSonOfLars reviewed (commented):
@blakepettersson I've just fixed some tests, and updated the PR off of latest master. Also added some tests and ensured that we aren't changing any existing behaviour.

Let me know if theres anything I can do to help get this in - probably best to get it in before v3!

PaulSonOfLars commented:
Not sure what the context of the `http: server gave HTTP response to HTTPS client` error might be; I'm guessing it's because I've removed the short-circuit in the helm version selection, in favour of checking the available tags, rather than blindly returning a version we don't know exists.
Let me know if theres something to be done to fix the test, or if you'd like me to undo add back that shortcircuit.

josecsotomorales commented:
Feel like this PR should be merged?

blakepettersson commented:
@PaulSonOfLars I would say let's re-add the short circuit for now and we can look into removing it at some other time.

PaulSonOfLars commented:
> @PaulSonOfLars I would say let's re-add the short circuit for now and we can look into removing it at some other time.

@blakepettersson Sure. I've left a note explaining context for future users. MIght be of interest to before v3 comes out.

Not sure what the test failure is about now, looks like some codegen for a piece I haven't touched?

lanceatcollibra commented:
Anyone know the status of this PR? Or are we waiting for 3.0?

PaulSonOfLars commented:
@crenshaw-dev @blakepettersson friendly nudge on the above.

This will help unify semver resolution across all providers (both git and helm), creating a more consistent and predictable approach, as well as resolve the issue #21818.
In case the concern is breaking changes, I would hope that the push to v3 would be a good excuse to get this merged in and tested. What do you think?

blakepettersson commented:
@PaulSonOfLars I'm all for it, I just don't have enough powers to merge it. Will ping some others to see if they can.

flaviomoringa commented:
@crenshaw-dev could you please help in merging this? We're are really stuck with errors like: 
`Failed to load target state: failed to generate manifest for source 1 of 1: rpc error: code = Unknown desc = invalid revision '33.03.138': constraint Parser Error`

Thanks

IanMoroney commented:
I have also just come across this issue, and i can't deploy an important helm chart into argo because of it.
```
Failed to load target state: failed to generate manifest for source 2 of 2: rpc error: code = 
Unknown desc = invalid revision '4.0-*': improper constraint: 4.0.002
```

lanceatcollibra commented:
Still waiting on this 🥲

blakepettersson commented:
@PaulSonOfLars there's a failing unit test, otherwise we're good to go

PaulSonOfLars commented:
@blakepettersson Fixed the test, should be good to go. 
Here's to hoping you can reenable the automerge!

blakepettersson commented:
@PaulSonOfLars now it's done, thanks a lot!

jose-arcatar commented:
Is this now available for v3? or until a next release?

rouke-broersma commented:
This was just merged so it's not on v3.0.0

rouke-broersma commented:
@blakepettersson Would it be possible to have this backported to 3.0? We currently cannot upgrade because we can't determine the impact to our customers, we would have to check that all their charts are compliant with the strict semantic versioning from before the fixes in this PR.

agaudreault commented:
@rouke-broersma seems like the cherry-pick bot is not working. Can you open  a PR targeting the release-3.0 branch?

rouke-broersma commented:
@agaudreault here it is: https://github.com/argoproj/argo-cd/pull/23310

Let me know if that's correct or not, it seems all good to me.
